### PR TITLE
Update SSHAction

### PR DESCRIPTION
**Pull request by knazir**
SSHAction reads the private key file from secure storage instead of a file path. Also updated docs to reflect property name changes. "passphrase" may be more appropriate than password. 

JIRA:  https://issues.cask.co/browse/HYDRA-169

### CODE REVIEW

**Comment by knazir** [comment]
Manually testing now and build running: http://builds.cask.co/browse/HYP-DBT50-2

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java [hunk 1]
```diff
@@ -116,49 +124,52 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
...
+    private String passphrase;
 
     @Nullable
     @Description("Variable used to hold desired key name to be associated with script output in the instance " +
       "the script output contains information needed to be stored in the ActionContext for later use in the pipeline.")
+    @Macro
     private String output;
 
-    public SSHActionConfig() {
-      port = 22;
```
Comment (anwar6953) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java:144]: this isn't the value used by default?

File: core-plugins/docs/SSHAction-action.md [hunk 2]
```diff
@@ -21,9 +21,9 @@ Properties
 
 **user:** The username used to connect to host.
 
-**privateKeyFile:** The file path to Private key.
+**privateKeyFileLookupKey:** The file path to Private key.
```
Comment (bdmogal) [inline core-plugins/docs/SSHAction-action.md]: nit: lowercase p

Comment (bdmogal) [inline core-plugins/docs/SSHAction-action.md]: Also, can this parameter be named better? Seems very verbose for something that's exposed to users. What does it do? Also seems like it could use some more detailed docs.

File: core-plugins/docs/SSHAction-action.md [hunk 3]
```diff
@@ -21,9 +21,9 @@ Properties
...
 
-**password:** The password associated with private key.
+**passphrase:** The password associated with private key.
```
Comment (bdmogal) [inline core-plugins/docs/SSHAction-action.md]: Is this related to the above property? If so, should mention the relationship.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java (hunk 1 above)
Comment (knazir) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java:144]: would this constructor ever be called? o-o

Comment (anwar6953) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java:144]: I think so.
But you're saying that the default value of 22 is not being set... so... not sure.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java [hunk 4]
```diff
@@ -61,12 +62,19 @@ public SSHAction(SSHActionConfig config) {
...
+
     Connection connection = new Connection(config.host);
     try {
       connection.connect();
 
-      String password = Strings.isNullOrEmpty(config.password) ? null : config.password;
-      if (!connection.authenticateWithPublicKey(config.user, new File(config.privateKeyFile), password)) {
+      // get private key from secure store using user-specified key
+      char[] privateKey =
+        new String(context.getSecureData(context.getNamespace(), config.privateKeyFileLookupKey).get()).toCharArray();
```
Comment (bdmogal) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java]: Should specify charset while converting from `byte[]` to `String`. Use `Bytes.toString(...)`.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java (hunk 1 above)
Comment (anwar6953) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java:144]: Seems like other code depends on it...
https://github.com/caskdata/hydrator-plugins/blob/develop/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/EmailAction.java#L103-L108

Note: regex to find all default constructors (intellij has regex search):
`public\ \w+\(\)\ \{`

Comment (anwar6953) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java:144]: What was the symptom observed in this JIRA?
https://issues.cask.co/browse/HYDRA-354
NPE? Or some other error?

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java [hunk 5]
```diff
@@ -116,49 +124,52 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
...
 
     @Nullable
-    @Description("Password associated with private key")
-    private String password;
+    @Description("Passphrase associated with private key")
+    @Macro
+    private String passphrase;
 
     @Nullable
     @Description("Variable used to hold desired key name to be associated with script output in the instance " +
```
Comment (bdmogal) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java]: Unrelated to your PR, but there are a couple of problems with this description:
1. Its hard to understand because of the long sentence.
2. It uses terms like `ActionContext`, which are very code specific. 

What does it do? Also, is it too internal to be exposed to users?

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java (hunk 1 above)
Comment (knazir) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java:144]: Not sure what I did to be honest, can't reproduce the behavior. I may have just used " " a few times. Will update the PR, thanks.

File: core-plugins/docs/SSHAction-action.md (hunk 2 above)
Comment (knazir) [inline core-plugins/docs/SSHAction-action.md]: Yeah fair point, we could try `privateKeyLookupKey`

File: core-plugins/docs/SSHAction-action.md (hunk 3 above)
Comment (knazir) [inline core-plugins/docs/SSHAction-action.md]: Yes it is, will update that.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java (hunk 5 above)
Comment (knazir) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java]: From my understanding it's meant to be exposed to users to (almost exclusively) substitute plugin properties with a macro `${output}` where "output" is the key they specified. I agree it could use some work. Will try to update it.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java [hunk 6]
```diff
@@ -116,49 +125,57 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
...
 
     @Nullable
-    @Description("Variable used to hold desired key name to be associated with script output in the instance " +
-      "the script output contains information needed to be stored in the ActionContext for later use in the pipeline.")
-    private String output;
+    @Description("The key used to store the output of the command run by the action. Plugins that run at later " +
+      "stages in the pipeline can retrieve the command's output using this key through macro substitution " +
+      "${outputKey} where \"outputKey\" is the key specified.")
+    @Macro
+    private String outputKey;
```
Comment (knazir) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java:150]: Since it's nullable, should a default key be given if none is provided like "sshOutput" or something?

File: core-plugins/docs/SSHAction-action.md [hunk 7]
```diff
@@ -21,15 +21,16 @@ Properties
 
 **user:** The username used to connect to host.
 
-**privateKeyFile:** The file path to Private key.
+**privateKeyLookupKey:** Key used to lookup private key from secure storage.
```
Comment (bdmogal) [inline core-plugins/docs/SSHAction-action.md]: Now it has 2 keys :) Actually, is it mandatory that this will always be a macro? If someone doesn't care about security, couldn't they just dump their private key in a textarea? I think we can just call it `privateKey`. The description can be something like - "The private key to be used to perform the secure shell action. Users can also specify a key that will be looked up from the secure key management store in CDAP". However, in that case you'll need a way to distinguish between whether its a secure key or just text.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java [hunk 8]
```diff
@@ -61,12 +62,20 @@ public SSHAction(SSHActionConfig config) {
...
 
-      String password = Strings.isNullOrEmpty(config.password) ? null : config.password;
-      if (!connection.authenticateWithPublicKey(config.user, new File(config.privateKeyFile), password)) {
+      // get private key from secure store using user-specified key
+      char[] privateKey =
+        Bytes.toString(context.getSecureData(context.getNamespace(), config.privateKeyLookupKey).get())
+          .toCharArray();
+      String passphrase = Strings.isNullOrEmpty(config.passphrase) ? "" : config.passphrase;
+
+      if (!connection.authenticateWithPublicKey(config.user, privateKey, passphrase)) {
```
Comment (bdmogal) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java]: so the method says authenticate with public key, but we ask for a private key?

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java (hunk 6 above)
Comment (bdmogal) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java:150]: Yes, and it should be specified in the description.

File: core-plugins/docs/SSHAction-action.md (hunk 7 above)
Comment (knazir) [inline core-plugins/docs/SSHAction-action.md]: That's a good point, I think I got too caught up in secure storage/macros.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java (hunk 8 above)
Comment (knazir) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java]: Right, it's a part of the ganymed2 SSH library. It's very poorly named...

Comment (bdmogal) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java]: but does it work with public keys? If so we should accept a public key.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java [hunk 9]
```diff
@@ -116,49 +114,59 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
...
-    private String output;
+    @Description("The key used to store the output of the command run by the action. Plugins that run at later " +
+      "stages in the pipeline can retrieve the command's output using this key through macro substitution " +
+      "${outputKey} where \"outputKey\" is the key specified. Defaults to \"sshOutput\".")
+    @Macro
+    private String outputKey;
 
     public SSHActionConfig() {
       port = 22;
+      outputKey = "sshOutput";
```
Comment (gsps1) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java]: use `this.outputKey`

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java [hunk 10]
```diff
@@ -116,49 +114,59 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
...
-      this.output = output;
+      this.privateKey = privateKeyFileLookupKey;
+      this.passphrase = passphrase;
+      this.outputKey = outputKey;
     }
 
     public void validate() {
       //check that port is not negative
-      if (port < 0) {
+      if (!containsMacro("port") && port < 0) {
```
Comment (gsps1) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java]: when you are calling  validate during runtime. `containsMacro("port")` will still be true even and you wouldn't check port if its negative or not.

Comment (knazir) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java]: Oh, I thought that we went with an empty set at runtime. Just checked PluginInstantiator and that's not the case.. will fix. Should the same check `if (port < 0)` just be copied to the run method @gsps1?

File: core-plugins/docs/SSHAction-action.md [hunk 11]
```diff
@@ -21,15 +21,17 @@ Properties
 
 **user:** The username used to connect to host.
 
-**privateKeyFile:** The file path to Private key.
+**privateKey:** The private key to be used to perform the secure shell action. Users can also specify a macro that will
```
Comment (gsps1) [inline core-plugins/docs/SSHAction-action.md:24]: it might be better to say `Users can store the private key in CDAP's secure store and use macro to read the private key from the secure store using ...` ask @awholegunch for review

**Comment by gsps1** [comment]
btw though unrelated to your changes, i noticed we are not closing the session, it's important to do so.  can you make that change as well?

Edit : seems like connection close closes the session too.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java (hunk 1 above)
Comment (gsps1) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java:144]: @knazir so we use reflection to instantiate and set values. but internally it will call default constructor before trying to set values for the variables through reflection. so say the port was not provided in the config, due to this default constructor being called, this config will have a value of `22` for port and whatever you set inside.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java (hunk 8 above)
Comment (knazir) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java]: Nope, it's just poorly named:

@param pemPrivateKey A <code>char[]</code> containing a DSA or RSA private key of the user in OpenSSH key format (PEM, you can't miss the "-----BEGIN DSA PRIVATE KEY-----" or "-----BEGIN RSA PRIVATE KEY-----" tag). The char array may contain linebreaks/linefeeds.

```
public synchronized boolean authenticateWithPublicKey(String user, char[] pemPrivateKey, String password);
```

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java [hunk 12]
```diff
@@ -95,68 +96,82 @@ public void run(final ActionContext context) throws Exception {
...
-      this.password = password;
-      this.output = output;
+      this.privateKey = privateKeyFileLookupKey;
+      this.passphrase = passphrase;
+      this.outputKey = outputKey;
+    }
+
+    public void validateConfigureTime() {
+      //check that port is not negative
+      if (!containsMacro("port") && port < 0) {
```
Comment (gsps1) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java]: i'd say just have validate only do `if (port < 0)` and in configurePipeline call validate if `config.containsMacro("port")`

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java [hunk 13]
```diff
@@ -95,65 +96,74 @@ public void run(final ActionContext context) throws Exception {
...
       }
 
-      if (config.output != null) {
-        //removes the carriage return at the end of the line
-        out = out.endsWith("\n") ? out.substring(0, out.length() - 1) : out;
-
-        context.getArguments().set(config.output, out);
-      }
+      //removes the carriage return at the end of the line
+      out = out.endsWith("\n") ? out.substring(0, out.length() - 1) : out;
```
Comment (gsps1) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java:100]: @sagarkapare should we handle multiple carriage return at the end ?

**Comment by knazir** [comment]
Manual testing looked good. Hydrator plugins tests are still broken so can't get a green build...

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java (hunk 13 above)
Comment (gsps1) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java:100]: spoke with sagar, we decided that's not needed

**Comment by gsps1** [comment]
code changes LGTM, please squash before merging.
